ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+   <Dependency wfid="1"/>
+   <Schedule/>
+  </Item>
+ </Workflow>
+
+Item 2 has a state of "reqd", so it should be evaluated now.  Item 2 has a dependency on item 1, so that must be
+evaluated first.  This is acheived by calling MyEclProcess::perform() on the object that was previously created
+from the workunit dll, passing in wfid = 1.  That will execute the following code::
+
+    switch (wfid) {
+        case 1U:
+            if (!gctx->isResult("searchname",4294967295U)) {
+                ctx->setResultString("searchname",4294967295U,5U,"Smith");
+            }
+            break;
+        break;
+    }
+
+This checks if a value has been provided for the input parameter, and if not assigns a default value of "Smith".
+The function returns and control returns to the workflow engine.  With the dependencies for wfid 2 now satisfied,
+the generated code for that workflow id is now executed::
+
+    switch (wfid) {
+        case 2U: {
+            ctx->executeGraph("graph1",false,0,NULL);
+            ctx->setResultString(0,1U,5U,"Done!");
+        }
+        break;
+    }
+    
+Most of the work for this workflow item involves executing graph1 (by calling back into the eclagent/roxie).  However
+it also directly sets another result.  This is fairly typical - the code inside MyProcess::perform often combines
+evaluating scalar results and temporary expressions, executing graphs, and calling functions
+that cannot (currently) be called inside a graph (e.g., those involving superfile transactions).
+
+Once all the required workflow items are executed the workunit is marked as completed.  Alternatively, if there are
+workflow items which are waiting to be triggered by an event the workunit will be passed to the scheduler
+which will keep monitoring for events.
+
+There are various specialised types of workflow items - e.g., sequential, wait, independent, but they all follow the same basic approach of
+executing dependencies and then executing that particular item.  
+
+Most of the interesting work in a ECL query is done within a graph.  The call ctx->executeGraph will either execute the graph locally
+(in the case of hthor and roxie), or add the workunit onto a queue (for Thor).  Whichever happens, control will pass to that engine...
+
+Executing a graph
+=================
+
+All the engines (roxie, hThor, Thor) execute graphs in a very similar way.  The main differences are that hThor and
+Thor execute a sub graph at a time, while roxie executes a complete graph as one.  Roxie is also optimized to minimize
+the overhead of executing a query - since the same query tends to be run multiple times - which means that it creates a
+graph of factory objects and those are used to create the activities.  The core details are the same for each of them though.
+
+Details of the graph structure
+------------------------------
+
+A reminder of the structure of the graph with the xml for the graph definition in our example.  Each graph (e.g., graph1)
```

Comment:
"A reminder of the structure of the graph with the xml for the graph definition in our example." I would re-word/structure this sentence.

*N.B,* In American English a comma should follow e.g. For example:
Female marsupials (e.g., kangaroos, opossums) have a pouch.
In British English no comma should follow e.g. For example:

Found this at http://en.wiktionary.org/wiki/e.g.